cartographer/mapping/internal/2d/pose_graph_2d.cc

```diff
@@ -480,6 +482,25 @@ void PoseGraph2D::HandleWorkQueue(
     num_nodes_since_last_loop_closure_ = 0;
   }
 
+  // Update the gauges that count the current number of constraints.
+  double inter_constraints_same_trajectory = 0;
+  double inter_constraints_different_trajectory = 0;
+  for (const auto& constraint : constraints()) {
```

Comment:
I think you need to hold the mutex while reading constraints, so move this block inside the block above.

Reply:
`constraints()` uses a mutex: https://github.com/googlecartographer/cartographer/blob/7d69ca9fa0393a0be2ae6dd5781d7a8ee9f4dcb1/cartographer/mapping/internal/2d/pose_graph_2d.cc#L946

Reply:
Then it is still preferable not to drop and take the same mutex but iterate over data_ instead while holding it.